### DISCUSSION TITLE
Match network layers together when training images that have more than 3 channels ?

### DISCUSSION THREAD

**Discussion by hakespear** [ultralytics/yolov5#7479] [Q&A] [answered]
Hi everyone,

I'd like to train a yolov5s network on several datasets for single class object detection. All of them are .tif images with 3 or more channels. I did successfully train the proposed network architecture on 3 channels images using the default configuration, but I couldn't do so for 4 channels for instance.

By looking around I found that the number of input channels can be specified. By adding `ch: 4` in dataset.yaml (as well as `nc: 1` because of single-class detection), and by specifying "ch=4" in both the # Model
```
    # Model
    check_suffix(weights, '.pt')  # check weights
    pretrained = weights.endswith('.pt')
    if pretrained:
        with torch_distributed_zero_first(LOCAL_RANK):
            weights = attempt_download(weights)  # download if not found locally
        ckpt = torch.load(weights, map_location='cpu')  # load checkpoint to CPU to avoid CUDA memory leak
        model = Model(cfg or ckpt['model'].yaml, ch=4, nc=nc, anchors=hyp.get('anchors')).to(device)  # create
        exclude = ['anchor'] if (cfg or hyp.get('anchors')) and not resume else []  # exclude keys
        csd = ckpt['model'].float().state_dict()  # checkpoint state_dict as FP32
        csd = intersect_dicts(csd, model.state_dict(), exclude=exclude)  # intersect
        model.load_state_dict(csd, strict=False)  # load
        LOGGER.info(f'Transferred {len(csd)}/{len(model.state_dict())} items from {weights}')  # report
    else:
        model = Model(cfg, ch=4, nc=nc, anchors=hyp.get('anchors')).to(device)  # create
```
and the # Model attributes
```
    # Model attributes
    nl = de_parallel(model).model[-1].nl  # number of detection layers (to scale hyps)
    hyp['box'] *= 4 / nl  # scale to layers
    hyp['cls'] *= nc / 80 * 4 / nl  # scale to classes and layers
    hyp['obj'] *= (imgsz / 640) ** 2 * 4 / nl  # scale to image size and layers
    hyp['label_smoothing'] = opt.label_smoothing
    model.nc = nc  # attach number of classes to model
    model.hyp = hyp  # attach hyperparameters to model
    model.class_weights = labels_to_class_weights(dataset.labels, nc).to(device) * nc  # attach class weights
    model.names = names
```
sections of train.py, I was able to change the very first conv layer such that it now effectively expects a 4 channels input. However, when training actually starts (the very first epoch just start to lauch at that point), I get this runtime error :
```
File "/home/.local/lib/python3.8/site-packages/torch/nn/modules/conv.py", line 444, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,
RuntimeError: Given groups=1, weight of size [32, 4, 6, 6], expected input[16, 3, 256, 256] to have 4 channels, but got 3 channels instead
Exception ignored in: <function StorageWeakRef.__del__ at 0x7f85b1ed20d0>
```

I understood that this error come from pytorch to tell that the input and output layers don't match at some point in the network architecture. Which somewhat disturb me since both the yolo network architecture might be 100% correct, and I checked my images and labels to not be corrupted and to have 4 channels.

Is there something I did wrong here, or did I forget something ? I'd appreciate any help on this.

**Comment by glenn-jocher** [comment, marked as answer]
@hakespear yes, model construction for arbitrary input channel counts is supported with the `ch:` key in model yamls as you saw, but this is not respected currently by the dataloaders, nor more broadly built into the repo. Instead we use automatic 3-ch scaling (i.e. black-white and 4-ch PNGs are converted to 3-ch).

If you'd like to attempt a PR to introduce n-channel training that would be great. We have a few suggestions for this around, i.e.:
https://github.com/ultralytics/yolov5/issues/6288#issuecomment-1096607430

Please see our ✅ [Contributing Guide](https://github.com/ultralytics/yolov5/blob/master/CONTRIBUTING.md) to get started.

**Reply by hakespear** [reply]
Much thanks for your response. That might explain the confusion between loading 4-ch images for 3-ch training and training directly on the 4 input channels.

If I get enough time to introduce n-channels training yes I'd like to do so. I could find a model architecture with `torch.hub.load('ultralytics/yolov5', 'yolov5s', channels=4, classes=1)` that only have two different arguments from the original yaml model (i.e. on the very first for channels and on the very last for classes) : 
```
                 from  n    params  module                                  arguments                     
  0                -1  1      4672  models.common.Conv                      [4, 32, 6, 2, 2]              
  1                -1  1     18560  models.common.Conv                      [32, 64, 3, 2]                
  2                -1  1     18816  models.common.C3                        [64, 64, 1]                   
  3                -1  1     73984  models.common.Conv                      [64, 128, 3, 2]               
  4                -1  2    115712  models.common.C3                        [128, 128, 2]                 
  5                -1  1    295424  models.common.Conv                      [128, 256, 3, 2]              
  6                -1  3    625152  models.common.C3                        [256, 256, 3]                 
  7                -1  1   1180672  models.common.Conv                      [256, 512, 3, 2]              
  8                -1  1   1182720  models.common.C3                        [512, 512, 1]                 
  9                -1  1    656896  models.common.SPPF                      [512, 512, 5]                 
 10                -1  1    131584  models.common.Conv                      [512, 256, 1, 1]              
 11                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 12           [-1, 6]  1         0  models.common.Concat                    [1]                           
 13                -1  1    361984  models.common.C3                        [512, 256, 1, False]          
 14                -1  1     33024  models.common.Conv                      [256, 128, 1, 1]              
 15                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 16           [-1, 4]  1         0  models.common.Concat                    [1]                           
 17                -1  1     90880  models.common.C3                        [256, 128, 1, False]          
 18                -1  1    147712  models.common.Conv                      [128, 128, 3, 2]              
 19          [-1, 14]  1         0  models.common.Concat                    [1]                           
 20                -1  1    296448  models.common.C3                        [256, 256, 1, False]          
 21                -1  1    590336  models.common.Conv                      [256, 256, 3, 2]              
 22          [-1, 10]  1         0  models.common.Concat                    [1]                           
 23                -1  1   1182720  models.common.C3                        [512, 512, 1, False]          
 24      [17, 20, 23]  1     16182  models.yolo.Detect                      [1, [[10, 13, 16, 30, 33, 23], [30, 61, 62, 45, 59, 119], [116, 90, 156, 198, 373, 326]], [128, 256, 512]]
YOLOv5s summary: 270 layers, 7023478 parameters, 7023478 gradients, 16.1 GFLOPs
```
I think rebuilding this architecture according to the modified extreme layers to make everything match would be the solution (without taking pretrain into account), but the yaml file seems to miss several layers arguments for simplicity.

Do you think it'll be easier to rebuild the entire network architecture with pytorch and then pass the number of channels parameter from there to dataloaders, or to just directly modify the yaml file's architecture section (plus the `ch:` and `nc:` keys) knowing some arguments are missing here ? (perhaps it's not necessary to change these ones)

**Reply by glenn-jocher** [reply]
@hakespear I don't understand your question.

**Reply by hakespear** [reply]
@glenn-jocher I mean there are arguments for Conv layers, C3 layers, and so on that are displayed in the model description files, about the backbone and head of the network architecture (models/yolov5s.yaml for instance) : 
```
anchors:
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32

# YOLOv5 v6.0 backbone
backbone:
  # [from, number, module, args]
  [[-1, 1, Conv, [64, 6, 2, 2]],  # 0-P1/2
   [-1, 1, Conv, [128, 3, 2]],  # 1-P2/4
   [-1, 3, C3, [128]],
   [-1, 1, Conv, [256, 3, 2]],  # 3-P3/8
   [-1, 6, C3, [256]],
   [-1, 1, Conv, [512, 3, 2]],  # 5-P4/16
   [-1, 9, C3, [512]],
   [-1, 1, Conv, [1024, 3, 2]],  # 7-P5/32
   [-1, 3, C3, [1024]],
   [-1, 1, SPPF, [1024, 5]],  # 9
  ]

# YOLOv5 v6.0 head
head:
  [[-1, 1, Conv, [512, 1, 1]],
   [-1, 1, nn.Upsample, [None, 2, 'nearest']],
   [[-1, 6], 1, Concat, [1]],  # cat backbone P4
   [-1, 3, C3, [512, False]],  # 13

   [-1, 1, Conv, [256, 1, 1]],
   [-1, 1, nn.Upsample, [None, 2, 'nearest']],
   [[-1, 4], 1, Concat, [1]],  # cat backbone P3
   [-1, 3, C3, [256, False]],  # 17 (P3/8-small)

   [-1, 1, Conv, [256, 3, 2]],
   [[-1, 14], 1, Concat, [1]],  # cat head P4
   [-1, 3, C3, [512, False]],  # 20 (P4/16-medium)

   [-1, 1, Conv, [512, 3, 2]],
   [[-1, 10], 1, Concat, [1]],  # cat head P5
   [-1, 3, C3, [1024, False]],  # 23 (P5/32-large)

   [[17, 20, 23], 1, Detect, [nc, anchors]],  # Detect(P3, P4, P5)
  ]
```
But I guess (I could be wrong) there are not as much displayed arguments as what can provide `torch.hub.load('ultralytics/yolov5', 'yolov5s', channels=4, classes=1)` from pytorch hub (previous post).
My hope is that modifying a bit the numbers from models/yolov5s.yaml to take n-channels into account could be enough for n-channels training (in addition to the `ch:` key). If I'm wrong here, then I think a new model should be entirely rebuilt from pytorch for every `n`. That'd be harder to implement since it'd require to go deeper in the code.

Thus my question is : do you think modifying models/yolov5s.yaml could technically be enough to implement n-channels training ? Or is it necessary to go deeper in the code ? (I should have talked about "possibility" rather than "ease" on my previous post).
Hoping to be more understandable, ofc please tell me if I'm still unclear.

**Reply by glenn-jocher** [reply]
@hakespear I already answered this below. Building n-channel models is trivial, that's already in place. The dataloaders (all of them, not just the training dataloader) and augmentation associated with n-channel data is not present in this repo.

> @hakespear yes, model construction for arbitrary input channel counts is supported with the `ch:` key in model yamls as you saw, but this is not respected currently by the dataloaders, nor more broadly built into the repo. Instead we use automatic 3-ch scaling (i.e. black-white and 4-ch PNGs are converted to 3-ch).
> 
> If you'd like to attempt a PR to introduce n-channel training that would be great. We have a few suggestions for this around, i.e.: [#6288 (comment)](https://github.com/ultralytics/yolov5/issues/6288#issuecomment-1096607430)
> 
> Please see our ✅ [Contributing Guide](https://github.com/ultralytics/yolov5/blob/master/CONTRIBUTING.md) to get started.

**Reply by hakespear** [reply]
Ok, so the key might be to change the dataloaders and adapt the repo to n-channel data. I somewhat thought building the n-channel models (which is quite trivial, I agree) would be enough, but got stuck with the `expected input[16, 3, 256, 256] to have 4 channels, but got 3 channels instead` error message. Thank you for having pointed out the issue, and thanks again for the suggestions.
